### ISSUE TITLE
notification like "Failed launch debugger for child process xxxx".

### ISSUE THREAD

**Issue by liheyi360** [microsoft/debugpy#712]
And i still  encounter the situation like #303 issue about multiple process debug, vscode pop up notification like "Failed launch debug for child process xxxx".
Sometimes, debugger can't acquire call stack info itself.
```python
import multiprocessing

def funcProcess(idx):
    return idx + 1

with multiprocessing.Pool() as myPool:
    ans = myPool.map(funcProcess, range(1000))
print(ans)

```

_Originally posted by @liheyi360 in https://github.com/microsoft/debugpy/issues/709#issuecomment-908139397_

**Comment by fabioz** [comment]
I still have to investigate this, but as a note, creating 1000 processes and subsequently 1000 debugs on VSCode may be stretching the limits (my guess is that the computer is getting too much to do and the timeouts end up reaching their limits).

Do you get this in a real-world use case?

**Comment by liheyi360** [comment]
I use this code to do some heavy work. And I think function Pool can't make 1000 processes, process number is related to cpu core. Python multiprocess module really attemp to creating 1000 processes? 
```python
import multiprocessing as mp
from os import getpid

def func(idx):
     # record process id, just some file created
    obj = open('{}.log'.format(getpid()), 'w')
    obj.close()
    print('{}\n'.format(getpid()))
    return idx+1

class myClass:
    def __init__(self):
        self.data = []

    def test(self):
        with mp.Pool(mp.cpu_count()) as mypool:
            ans = mypool.map(func, range(100))
        self.data = ans

a = myClass()
a.test()
```

**Comment by fabioz** [comment]
You're right, it should cap on your cpu count for simultaneous processes (then it'll start to reuse processes)... If you use less cores so that it doesn't stay at 100% cpu utilization for all cores, does it still have that issue for you (say use cpu_count()/2)?

**Comment by liheyi360** [comment]
I follow your advice, notification still happen. 
![Screenshot from 2021-08-31 21-07-22](https://user-images.githubusercontent.com/44501220/131508277-e3e4e8f2-f918-4e7e-beb0-e55ee38230d1.png)
[log.zip](https://github.com/microsoft/debugpy/files/7084113/log.zip)
 
```python
import multiprocessing as mp
from os import getpid

def initFunc():
    obj = open('{}.log'.format(getpid()), 'w')
    obj.close()

def func(idx):
    return idx+1

class myClass:
    def __init__(self):
        self.data = []

    def test(self):
        with mp.Pool(int(mp.cpu_count()/2), initFunc) as mypool:
            ans = mypool.map(func, range(10000))
        self.data = ans

a = myClass()
a.test()
```

**Comment by int19h** [comment]
Does it make any difference if you do `mp.set_start_method('spawn')`?

**Comment by int19h** [comment]
The subprocesses should be paused until a client connects to them. But, for some reason, they do indeed exit early - in fact, most of them exit before the client even gets to request attach!

It's not clear to me why this is happening. The logs don't show anything unusual. Also, so far as I can tell, this repros on Linux (regardless of start method), but not on Windows.

**Comment by liheyi360** [comment]
It's a misunderstanding.
I just set the breakpoint at the line where sub process has ended. Debugger can't launch non-existing process.
Code just work well, though notification is just a  little annoying but needed.

**Comment by int19h** [comment]
Can you clarify? It still looks like a bug to me in my local repros. When *not* debugging, everything works when expected. But when debugging, only a few processes spawn successfully; the rest exit before they even get to `initFunc()` (one can see that execution didn't get there by looking at those `<pid>.log` files the function creates). This is the weird part - when a subprocess is spawned, it should not do anything until some client connects to it. So it's unclear how or why it's exiting.

(My suspicion is that this has something to do with `multiprocessing` itself - maybe there's some timeout there? But I couldn't find anything obvious from a quick glance at the source.)

**Comment by liheyi360** [comment]
The last comment is confusing and not verified, just forget it.
And I'm sorry I have no enough time and knowledge to discuss in depth.
Thank you for your help.

**Comment by fabioz** [comment]
I just investigated this... apparently, adding a bit of timeout before the pool is finished makes the error go away
i.e.: something as:
```
with multiprocessing.Pool() as myPool:
    self.data = myPool.map(funcProcess, range(1000))
    time.sleep(1)
```

Alternatively, just making more happen in the pool (say, change `range(1000)` to `range(100000)`) also makes the error go away.

So, what seems to be happening is:

- The pool is created 
- Functions are scheduled to run on the pool
- A few processes are created and start running the functions scheduled
- When a process is started, a notification is sent to vscode to connect to it
- The process/pool finishes and the processes are killed
- VSCode didn't connect fast enough to debugpy for some of the processes which are now killed and complains about it (either with a `Failed to launch debugger for child process: <pid>` or with the `Server disconnected unexpectedly` message depending on to which point it got to into the connection process).

As a note, the execution seems to work fine (i.e.: the `self.data` is properly assigned with all expected values) when that message is shown, so, the main issue is just that those messages are being shown because the connection to the client wasn't completed before the process was killed.

Given that this is in the connection management layer (`vscode <-> debugpy`), would you like to take a look at that @int19h?

**Comment by int19h** [comment]
Thing is, child processes aren't supposed to start running anything until there's a debugger connection to them that has gone through the entire initialization stage (otherwise breakpoints might be skipped etc). So either the pool killing processes due to some kind of timeout, before they even had a chance to run anything; or there's some bug in how subprocesses are resumed.

**Comment by fabioz** [comment]
> Thing is, child processes aren't supposed to start running anything until there's a debugger connection to them that has gone through the entire initialization stage (otherwise breakpoints might be skipped etc)

We're on the same page there.

> So either the pool killing processes due to some kind of timeout, before they even had a chance to run anything

Exactly, but not due to some timeout and as its regular operation when everything scheduled to run already finished.

i.e.: The issue is that the multiprocessing pool would start 8 processes, starts 4 and starts sending the methods to be executed while the other ones are still being initialized, then, all the methods end up finishing in those first 4 processes and the remaining 4 which weren't ready to run are killed when the multiprocessing pool context manager exits in that example (while the debugger is still in the connection phase for those processes).

**Comment by ShorelLee** [comment]
I also encountered this problem, but my problem is, I hit a breakpoint in the program, but the program has returned before the breakpoint, so this problem occurs.

**Comment by david-waterworth** [comment]
I just encountered this as well, 3/4 of the way through a machine learning task vscode started displaying message boxes "Failed to launch debugger for child process" and the main script terminated with no exceptions printed to the console.

**Comment by int19h** [comment]
Note that the in the original issue, the computation itself completes successfully with the expected result - the error messages are essentially spurious (they're technically correct, just irrelevant).

If you're actually seeing *different results* with and without debugger, can you please file a separate issue?

**Comment by tierriminator** [comment]
I am having the same issue and for me it is very invasive since the `Server disconnected unexpectedly` message needs to be actively cancelled. I am using `Pool.map` from `multiprocessing` rather frequently, which makes it effectively impossible for me to debug anything that isn't right at the beginning of my program.

**Comment by BraveDrXuTF** [comment]
Similar problem, mine is "invalid message session is already started"

**Comment by BraveDrXuTF** [comment]
> Similar problem, mine is "invalid message session is already started"

How to fix it?
